### PR TITLE
feat: `make_selectpaulirot_to_phase_gradient_decomp` is capture compatible

### PR DESCRIPTION
**Pull request by andrijapau**
**Context:**

This decomp rule factory needs to be capture compatible.

**Benefits:**

This unlocks a necessary step to make it `qjit(capture=True)` compatible.

**Possible Drawbacks:** None identified.

[sc-120723]

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -21,6 +21,9 @@
 
 <h3>Labs: a place for unified and rapid prototyping of research software 🧪</h3>
 
+* Updated the `make_selectpaulirot_to_phase_gradient_decomp` decomposition rule factory to be compatible with program capture.
```
Comment (dwierichs) [inline doc/releases/changelog-dev.md]: Maybe worth batching this with the changelog entry from #9481 ?

File: pennylane/labs/tests/transforms/test_decomp_selectpaulirot_phase_gradient.py [hunk 2]
```diff
@@ -203,3 +204,69 @@ def circuit(angles, in_state):
...
+
+            @DecomposeInterpreter(
+                gate_set=gate_set, fixed_decomps={qp.SelectPauliRot: custom_decomp}
+            )
+            def f(angles):
+                qp.SelectPauliRot(angles, control_wires=control_wires, target_wire=target_wire)
+                return qp.state()
+
+            cjaxpr = jax.make_jaxpr(f)(angles)
+            print(cjaxpr)
```
Comment (dwierichs) [inline pennylane/labs/tests/transforms/test_decomp_selectpaulirot_phase_gradient.py]:
```suggestion
```

File: pennylane/labs/tests/transforms/test_decomp_selectpaulirot_phase_gradient.py [hunk 3]
```diff
@@ -203,3 +204,69 @@ def circuit(angles, in_state):
...
+            custom_decomp = make_selectpaulirot_to_phase_gradient_decomp(
+                angle_wires, phase_grad_wires, work_wires
+            )
+
+            gate_set = {
+                "QROM",
+                "Adjoint(QROM)",
+                "SemiAdder",
+                "CNOT",
+                "Adjoint(CNOT)",
```
Comment (dwierichs) [inline pennylane/labs/tests/transforms/test_decomp_selectpaulirot_phase_gradient.py]: Not really the scope of this PR, but do we need the `Adjoint(CNOT)` and `Adjoint(PauliX)` here? Would be nicer if not.

File: pennylane/labs/tests/transforms/test_decomp_selectpaulirot_phase_gradient.py [hunk 4]
```diff
@@ -203,3 +204,69 @@ def circuit(angles, in_state):
...
+                "PauliX",
+                "Adjoint(PauliX)",
+                "GlobalPhase",
+            }
+
+            @DecomposeInterpreter(
+                gate_set=gate_set, fixed_decomps={qp.SelectPauliRot: custom_decomp}
+            )
+            def f(angles):
```
Comment (dwierichs) [inline pennylane/labs/tests/transforms/test_decomp_selectpaulirot_phase_gradient.py:255]: Would there be any benefit in testing with wires as an argument here, to make them tracers? 🤔

**Review by dwierichs — CHANGES_REQUESTED** [review]
Happy to see that there are no major changes needed! Thanks @andrijapau 🎉 

Just some minor stuff, basically ready to approve.

File: pennylane/labs/tests/transforms/test_decomp_selectpaulirot_phase_gradient.py (hunk 4 above)
Comment (andrijapau) [reply pennylane/labs/tests/transforms/test_decomp_selectpaulirot_phase_gradient.py:255]: That's a good idea! I can try it out and see if it's possible. :)

File: pennylane/labs/tests/transforms/test_decomp_selectpaulirot_phase_gradient.py (hunk 3 above)
Comment (andrijapau) [reply pennylane/labs/tests/transforms/test_decomp_selectpaulirot_phase_gradient.py]: Yeah, I was getting errors if I didn't include them in the gate set. 😢 

I'll try and make a MWE and report it.

File: pennylane/labs/tests/transforms/test_decomp_selectpaulirot_phase_gradient.py (hunk 2 above)
Comment (andrijapau) [reply pennylane/labs/tests/transforms/test_decomp_selectpaulirot_phase_gradient.py]: Whoops. Thanks!

File: doc/releases/changelog-dev.md (hunk 1 above)
Comment (andrijapau) [reply doc/releases/changelog-dev.md]: Yeah, this was more of a placeholder until the #9481 PR goes in. Once it does I can adjust it! 😄

File: pennylane/labs/tests/transforms/test_decomp_selectpaulirot_phase_gradient.py (hunk 3 above)
Comment (andrijapau) [reply pennylane/labs/tests/transforms/test_decomp_selectpaulirot_phase_gradient.py]: Ah, so I think it's because I never specified the uncompute function here,
```python
    def inner_cob():
        return change_op_basis(compute_fn, target_fn)
```
In this case, the uncompute function will be determined by taking the adjoint of the compute_fn which will result in some `Adjoint(CNOT)` and `Adjoint(X)` gates. Perhaps because we're not executing the code, no simplification occurs resulting in the graph decomp system seeing these ops and requiring them to be in the gate set.

If I adjust it to be,
```python
    def inner_cob():
        return change_op_basis(compute_fn, target_fn, compute_fn)
```
and remove those ops from the gate set, everything works. 🤔

File: pennylane/labs/tests/transforms/test_decomp_selectpaulirot_phase_gradient.py (hunk 4 above)
Comment (andrijapau) [reply pennylane/labs/tests/transforms/test_decomp_selectpaulirot_phase_gradient.py:255]: Hmm, I'm getting some tracer errors. Is this something we fundamentally want to support? 😓

File: pennylane/labs/tests/transforms/test_decomp_selectpaulirot_phase_gradient.py (hunk 3 above)
Comment (dwierichs) [reply pennylane/labs/tests/transforms/test_decomp_selectpaulirot_phase_gradient.py]: I see 🤔 I would expect that even if `change_op_basis` produces such ops, it should be okay to not have them in the gate set because `Adjoint(CNOT)` should be decomposed to `CNOT`.

File: pennylane/labs/tests/transforms/test_decomp_selectpaulirot_phase_gradient.py (hunk 4 above)
Comment (dwierichs) [reply pennylane/labs/tests/transforms/test_decomp_selectpaulirot_phase_gradient.py:255]: mhm I think it would be nice but for XAS I don't see us tracing those particular wires necessarily (it would be nicer for scalability but should be okay to not have this for now). For `RZ` we want traced target wires :)

**Review by dwierichs — APPROVED** [review]
Thank you @andrijapau  🎉 

Would be nice to understand why `Adjoint(CNOT)` needs to be in the gate set and is not decomposed to `CNOT` itself, even if the `uncompute_fn` is not provided. That is somewhat tangential to this PR though :)

I would be fine deferring the wire tracing for now, it's not guaranteed that we will need this. May I ask you to document what you found, and maybe distill a downstream bug report/feature request from that? :) Thanks!

**Comment by andrijapau** [comment]
> Thank you @andrijapau 🎉
> 
> Would be nice to understand why `Adjoint(CNOT)` needs to be in the gate set and is not decomposed to `CNOT` itself, even if the `uncompute_fn` is not provided. That is somewhat tangential to this PR though :)
> 
> I would be fine deferring the wire tracing for now, it's not guaranteed that we will need this. May I ask you to document what you found, and maybe distill a downstream bug report/feature request from that? :) Thanks!

Sounds good! I'll do that and link it here. 😄

File: pennylane/labs/tests/transforms/test_decomp_selectpaulirot_phase_gradient.py (hunk 3 above)
Comment (andrijapau) [reply pennylane/labs/tests/transforms/test_decomp_selectpaulirot_phase_gradient.py]: Agreed. Let me see if I can make a MWE to report!

**Review by astralcai — APPROVED** [review]

File: doc/releases/changelog-dev.md (hunk 1 above)
Comment (andrijapau) [reply doc/releases/changelog-dev.md]: Done!

File: pennylane/labs/tests/transforms/test_decomp_selectpaulirot_phase_gradient.py (hunk 3 above)
Comment (andrijapau) [reply pennylane/labs/tests/transforms/test_decomp_selectpaulirot_phase_gradient.py]: Actually, I don't think this is an issue with the decomposition system, I think it's just a quirk with my testing strategy.

Consider this circuit,
```python
import pennylane as qp
import jax
from pennylane.ops.op_math import change_op_basis
from pennylane.tape.plxpr_conversion import CollectOpsandMeas
from pennylane.transforms.decompose import DecomposeInterpreter

qp.capture.enable()
qp.decomposition.enable_graph()

def compute_fn():
    qp.PauliX(0)
    qp.CNOT([0, 1])

def target_fn():
    qp.RZ(0.5, 1)

@qp.register_resources({qp.PauliX: 2, qp.CNOT: 2, qp.RZ: 1})
def myop_to_cob(wires):
    change_op_basis(compute_fn, target_fn)   # uncompute defaults to adjoint(compute_fn)

class MyOp(qp.operation.Operator):
    num_wires = 1
    resource_keys = set()

@DecomposeInterpreter(gate_set={"PauliX", "CNOT", "RZ"}, fixed_decomps={MyOp: myop_to_cob})
def f():
    MyOp(0)
    return qp.state()
```
the `jaxpr` is,
```python
>>> jaxpr = jax.make_jaxpr(f)()
>>> print(jaxpr)
{ lambda ; . let
    _:i64[1] = broadcast_in_dim[
      broadcast_dimensions=()
      shape=(1,)
      sharding=None
    ] 0:i64[]
    _:AbstractOperator() = PauliX[n_wires=1] 0:i64[]
    _:AbstractOperator() = CNOT[n_wires=2] 0:i64[] 1:i64[]
    _:AbstractOperator() = RZ[n_wires=1] 0.5:f64[] 1:i64[]
    adjoint_transform[
      jaxpr={ lambda ; . let
          _:AbstractOperator() = PauliX[n_wires=1] 0:i64[]
          _:AbstractOperator() = CNOT[n_wires=2] 0:i64[] 1:i64[]
        in () }
      lazy=True
      n_consts=0
    ]
    a:AbstractMeasurement(n_wires=0) = state_wires
  in (a,) }
```

Now, because the `adjoint` in COB is lazy, when you try to collect all of the ops and meas from the jaxpr, they won't be simplified:
```python
>>> collector = CollectOpsandMeas()
>>> collector.eval(jaxpr.jaxpr, jaxpr.consts)
>>> print([op.name for op in collector.state["ops"]])
['PauliX', 'CNOT', 'RZ', 'Adjoint(CNOT)', 'Adjoint(PauliX)']
```
Because my testing strategy is the following,
```python
assert op_names.issubset(
                gate_set
            ), f"Following ops are present but not in gateset: {op_names - gate_set}"
```
you'd need to add these `Adjoint(...)` operators in the gate set in order to pass. 🤦🏼

**Comment by andrijapau** [comment]
> > Thank you @andrijapau 🎉
> > Would be nice to understand why `Adjoint(CNOT)` needs to be in the gate set and is not decomposed to `CNOT` itself, even if the `uncompute_fn` is not provided. That is somewhat tangential to this PR though :)
> > I would be fine deferring the wire tracing for now, it's not guaranteed that we will need this. May I ask you to document what you found, and maybe distill a downstream bug report/feature request from that? :) Thanks!
> 
> Sounds good! I'll do that and link it here. 😄

Hey @dwierichs ! I'm not convinced it's a bug anymore, please see my message [here](https://github.com/PennyLaneAI/pennylane/pull/9537#discussion_r3326535653).

File: pennylane/labs/tests/transforms/test_decomp_selectpaulirot_phase_gradient.py (hunk 3 above)
Comment (dwierichs) [reply pennylane/labs/tests/transforms/test_decomp_selectpaulirot_phase_gradient.py]: Ahhh okay I see! Then I'm fine with this, personally :)

**Comment by dwierichs** [comment]
Nice, thanks for following up on this @andrijapau! 
Did you document the wires tracing issue you hit as well? :)

**Comment by andrijapau** [comment]
> Nice, thanks for following up on this @andrijapau! Did you document the wires tracing issue you hit as well? :)

I've created a P2 [story](https://app.shortcut.com/xanaduai/story/121177/control-target-wires-are-traceable-with-custom-phase-gradient-selectpaulirot-decomposition) to track the effort, but I've been distracted with a [bug](https://github.com/PennyLaneAI/pennylane/pull/9552/changes) that arose during the investigation. 😄 

If it's okay with you, I'll merge this PR in and then add more details to the story as I continue my investigation.

**Comment by dwierichs** [comment]
>If it's okay with you, I'll merge this PR in and then add more details to the story as I continue my investigation.

Great, thanks @andrijapau :D Of course, go ahead 👌
